src/metrics/nom.rs

```diff
@@ -260,20 +279,23 @@ mod tests {
     }
 
     #[test]
-    fn c_nom() {
+    fn javascript_nom() {
         check_metrics!(
-            "int foo();
-
-             int foo() {
-                 return 0;
+            "function f(a, b) {
+                 function foo(a) {
+                     return a;
+                 }
+                 var bar = function (a, b) {return a + b};
+                 var bar1 = function (a) {return a};
```

Comment:
Why considering these differently than `foo`? They are all named functions.
We need to decide on our definition of what a closure is

Reply:
Because in the `foo` case we have a `FunctionDefinition`, while in the `bar` case we have a function _assigned_ to a variable. I've used this criteria to discriminate among functions and closures

Reply:
Perhaps we can use this kind of closures for `Javascript`:

```
var add = (function () {
  var counter = 0;
  return function () {counter += 1; return counter}
})();

add();
add();
add();
```

Taken from [here](https://www.w3schools.com/js/js_function_closures.asp)

Reply:
Reading more carefully the example, in JavaScript, a closure is a mechanism to make global variables as local variables of a function. Do we want to detect this mechanism as a closure?

Reply:
Basically, any JS function forms a closure.
From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures:
> A closure is the combination of a function bundled together (enclosed) with references to its surrounding state (the lexical environment). In other words, a closure gives you access to an outer function’s scope from an inner function. In JavaScript, closures are created every time a function is created, at function creation time.

Maybe we should totally change what we're doing, remove the "closure" concept and define a "inner function" concept instead?
@calixteman what do you think?

Reply:
Since the definition above describes what a closure is, I propose to maintain the word "closure" and change the implementation instead. Introducing a new word could be confusing. What do you think @marco-c and @calixteman?

Reply:
@Luni-4 in JavaScript any function matches that definition, so if we keep using the word "closure" we'd need to consider any function as a closure for the JS language.

Reply:
@marco-c So why have functions and closures been splitted in? So a function doesn't exist at all because this property: _In other words, a closure gives you access to an outer function’s scope from an inner function._  is always true?

In that case, let's count only functions, specifying that functions and closures are the same

Reply:
We initially considered a "closure" an "inner function", because in all languages except JS that works well enough. In JS instead any function is a closure, so that assumption is broken.
I wouldn't keep the same word ("closure") now, as I think there is often a logic difference between "top-level functions" and "inner functions".

We could do one of or a mix of:
- separate functions between "named functions" and "unnamed functions" (though often small inner functions have a name);
- separate functions between "top-level functions" and "inner functions" (though this doesn't work with a common pattern of JS like:
```
(function() {
    function ciao() {
        // Ciao
    }
})();
```

```
function ciao() {
    // Ciao
}
```
these two would be considered differently, but I would instead consider them the same).

I don't see a clear winner

Reply:
> Basically, any JS function forms a closure.
> From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures:
> 
> > A closure is the combination of a function bundled together (enclosed) with references to its surrounding state (the lexical environment). In other words, a closure gives you access to an outer function’s scope from an inner function. In JavaScript, closures are created every time a function is created, at function creation time.
> 
> Maybe we should totally change what we're doing, remove the "closure" concept and define a "inner function" concept instead?
> @calixteman what do you think?

I think we don't care so much about detail of implementation.
From my pov, there are 2 kind of functions: named ones and anonymous ones.
Named ones may have different patterns: `function name(...`, ` name = (...) => ...`, `name = (funtion(){ return function()...)())`.
Everything is a kind of "inner function"... so I'm not sure that it's relevant here.

Reply:
This would be the first option in my last comment:
> separate functions between "named functions" and "unnamed functions" (though often small inner functions have a name);

The only drawback is that some developers assign names to all functions, and then we'd lose the differentiation.

Reply:
To wrap it up, in *JS*:

- Named functions should be treated as functions
- Unnamed functions like `return function()...)()` as closures

Is this the kind of approach we want to follow to?